File: apps/lazuar-pay/src/Lazuar.Pay/Identity/Client/Bearer.cs
using Lazuar.Pay.Hosting;

namespace Lazuar.Pay.Identity.Client;

internal static class Bearer
{
    const string Prefix = "Bearer ";

    public static bool TryGet(HttpRequest request, out string authorization)
    {
        authorization = request.Headers.Authorization.ToString();
        if (string.IsNullOrWhiteSpace(authorization))
        {
            return false;
        }

        if (!authorization.StartsWith(Prefix, StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        return authorization.Length > Prefix.Length && !string.IsNullOrWhiteSpace(authorization[Prefix.Length..]);
    }

    public static string Token(string authorization) =>
        authorization.StartsWith(Prefix, StringComparison.OrdinalIgnoreCase)
            ? authorization[Prefix.Length..].Trim()
            : authorization.Trim();

    /// <summary>One machine key. Not a JWT, not Stripe/Hub <c>sk_</c>.</summary>
    public static bool IsMachineKey(string authorization) =>
        Token(authorization).StartsWith("lzr_sk_", StringComparison.Ordinal);

    /// <summary>
    /// Stripe/Hub <c>sk_</c> as Pay Authorization is the wrong family.
    /// Vault PUT still accepts <c>sk_test</c> in JSON. Zitadel PAT shapes are
    /// not inventoried; they fail at One if they are not <c>lzr_sk_</c> or JWT.
    /// </summary>
    public static IResult? RejectWrongFamily(string authorization)
    {
        var token = Token(authorization);
        if (token.StartsWith("sk_", StringComparison.Ordinal)
            || token.StartsWith("sk_test_", StringComparison.Ordinal)
            || token.StartsWith("sk_live_", StringComparison.Ordinal))
        {
            return PayErrors.Status(401, "Unauthorized", "Invalid bearer");
        }

        return null;
    }
}

File: apps/lazuar-api/src/identity/bearer.rs
//! Port of `Identity/Client/Bearer.cs`.

const PREFIX: &str = "Bearer ";

/// Authorization header present, `Bearer `-prefixed, non-empty payload.
pub fn try_get(auth_header: Option<&str>) -> Option<String> {
    let authorization = auth_header?.trim();
    if authorization.is_empty() || !authorization.to_ascii_uppercase().starts_with(&PREFIX.to_uppercase()) {
        return None;
    }
    let token = authorization[PREFIX.len()..].trim();
    (!token.is_empty()).then(|| authorization.to_string())
}

pub fn token(authorization: &str) -> &str {
    if authorization.len() > PREFIX.len()
        && authorization[..PREFIX.len()].eq_ignore_ascii_case(PREFIX)
    {
        authorization[PREFIX.len()..].trim()
    } else {
        authorization.trim()
    }
}

/// One machine key. Not a JWT, not Stripe/Hub `sk_`.
pub fn is_machine_key(authorization: &str) -> bool {
    token(authorization).starts_with("lzr_sk_")
}

/// Stripe/Hub `sk_` as Pay Authorization is the wrong family.
pub fn reject_wrong_family(authorization: &str) -> Option<crate::hosting::PayError> {
    let t = token(authorization);
    if t.starts_with("sk_") {
        Some(crate::hosting::PayError::unauthorized("Invalid bearer"))
    } else {
        None
    }
}
